### PR TITLE
Verify output files before treating a script as fresh

### PR DESCRIPTION
**Pull request by aomarks**
### Background

Before this PR, if the `.wireit/<script>/fingerprint` file matches the current fingerprint for a script, then we assume that script is fresh, and skip it.

This works great, unless another process modifies the output files independently of Wireit, in which case the script will be skipped even though it actually needed to run.

### This PR

Now, after running a script, we now list all files matching the `output` globs, and save a record of their metadata (name, mtime, size) into a file named `.wireit/<script>/manifest`.

Then when preparing to run a script, after noticing that the current fingerprint matches the `.wireit/<script>/fingerprint` file, we re-compute the manifest (in the same way as we did above), and compare it to the `.wireit/<script>/manifest` file. If they match, consider the script fresh, and skip. If they don't match, consider the script stale, log an info message, and act the same way as we would if the fingerprint didn't match.

Fixes https://github.com/google/wireit/issues/292
Fixes https://github.com/google/wireit/issues/280
Fixes https://github.com/google/wireit/issues/262
Fixes https://github.com/google/wireit/issues/150

### Other changes

- Previously, if `output` was undefined, we could still treat a script as fresh, because we assumed the output wouldn't have changed if the input files didn't change. We no longer assume this, so now it's necessary to specify `output` to get freshness skipping (can be empty array).

- Glob results are now always returned as absolute paths. This made it simpler to share the cached output glob results between manifest generation and cleaning.

- Fixed a test rig bug where `WIREIT_FAILURES` would get propagated to the test cases.

### CODE REVIEW

File: src/util/manifest.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  m: number | undefined;
+  /** Size in bytes. `undefined` for directories. */
+  s: number | undefined;
+}
+
+/**
+ * A JSON-serialized manifest of files.
+ */
+export type FileManifestString = string & {
+  __OutputManifestStringBrand__: never;
```
Comment (augustjk) [inline src/util/manifest.ts:34]: Super nit. Should it be `__FileManifestStringBrand__` to match the type name?

File: src/execution/one-shot.ts [hunk 2]
```diff
@@ -458,45 +502,86 @@ export class OneShotExecution extends BaseExecution<OneShotScriptConfig> {
...
+
+  /**
+   * Glob the output files for this script and cache them, but throw unless the
+   * script has not yet started running or been restored from cache.
+   */
+  #globOutputFilesBeforeRunning(): Promise<
+    Result<AbsoluteEntry[] | undefined>
+  > {
+    this.#ensureState('before-running');
+    return (this.#cachedOutputFilesBeforeRunning ??= this.#globOutputFiles());
```
Comment (augustjk) [inline src/execution/one-shot.ts:550]: Neat!

File: src/test/freshness.test.ts [hunk 3]
```diff
@@ -1749,4 +1838,107 @@ test(
   })
 );
 
+test(
+  'script is not fresh if output file is modified externally',
```
Comment (augustjk) [inline src/test/freshness.test.ts:1842]: Love this test!

**Review by augustjk — APPROVED** [review]
Can't wait for this to be released!

File: src/execution/one-shot.ts [hunk 4]
```diff
@@ -458,45 +502,86 @@ export class OneShotExecution extends BaseExecution<OneShotScriptConfig> {
...
+    Result<AbsoluteEntry[] | undefined>
+  >;
+
+  /**
+   * Glob the output files for this script and cache them, but throw unless the
+   * script has finished running or been restored from cache.
+   */
+  #globOutputFilesAfterRunning(): Promise<Result<AbsoluteEntry[] | undefined>> {
+    this.#ensureState('after-running');
+    return (this.#cachedOutputFilesAfterRunning ??= this.#globOutputFiles());
```
Comment (rictic) [inline src/execution/one-shot.ts:562]: elegant!

**Review by rictic — APPROVED** [review]

**Review by AndrewJakubowicz — COMMENTED** [review]
Awesome PR!

File: src/caching/github-actions-cache.ts [hunk 5]
```diff
@@ -147,18 +147,18 @@ export class GitHubActionsCache implements Cache {
...
     }
 
-    const absFiles = relativeFiles.map((rel) =>
-      pathlib.join(script.packageDir, rel.path)
-    );
     const tempDir = await makeTempDir(script);
     try {
-      const tarballPath = await this.#makeTarball([...absFiles], tempDir);
+      const tarballPath = await this.#makeTarball(
+        [...absFiles.map((file) => file.path)],
```
[the remark this answers is not in the document]
Comment (aomarks) [reply src/caching/github-actions-cache.ts:159]: Good point, done in https://github.com/google/wireit/pull/288

**Pull request merged**